lib/hourly_logger_rotator/patch/ruby_2_4.rb

```diff
@@ -24,6 +24,27 @@ def initialize(log = nil, shift_age: nil, shift_size: nil, shift_period_suffix:
         end
       end
 
+      def shift_log_period(period_end)
+        suffix = period_end.strftime(@shift_period_suffix)
+        age_file = "#{@filename}.#{suffix}"
+        if FileTest.exist?(age_file)
+          # try to avoid filename crash caused by Timestamp change.
+          idx = 0
+          # .99 can be overridden; avoid too much file search with 'loop do'
+          while idx < 100
+            idx += 1
+            age_file = "#{@filename}.#{suffix}.#{idx}"
+            break unless FileTest.exist?(age_file)
+          end
+        end
+        @dev.close rescue nil
+        File.rename("#{@filename}", age_file)
+        @dev = create_logfile(@filename)
+        HourlyLoggerRotator::Gzip.new(age_file).call if HourlyLoggerRotator.gzip
```

Comment:
I don't think it's good idea to put this stuff to `shift_log_period` because:
1. It's going to slow down the evaluation of some random process like web-request or other things that happen during the log shift period.
2. `HourlyLoggerRotator::LogsLifetime` should only happen once and not for each logger instance.

Reply:
how to fix it?

Reply:
I agree with the 'slow down' part